Feat/name convention-Name convention handling

AhmadNader319 opened the pull request:
Enhance Java code formatting with naming conventions

This pull request includes significant updates to the Java code formatting logic by integrating naming convention methods. The most important changes involve modifying the FormattingVisitor class to handle various Java constructs with consistent naming conventions and updating the NameConventionHandler class to provide these methods.

- Added NameConventionHandler #2  class with methods:
  - to_pascal_case: Converts class names to PascalCase.
  - to_camel_case: Converts method (including method's parameters) and variable names to camelCase.
  - to_full_case: Converts constant variables to FULL_CASE. (to perform)

- Updated FormattingVisitor class to use NameConventionHandler methods:
  - Class names are converted to PascalCase. (to create an instance 
  - Method names are converted to camelCase.
  - Parameter names are converted to camelCase.
  - Local variable names are converted to camelCase.
  - Identifiers are replaced with their camelCase equivalents if they exist in the replacement dictionaries.(local variables for now unified)

These enhancements ensure consistent naming conventions throughout the codebase, making it easier to read and maintain.

AhmadNader319 committed:
NameConventio class with methods to pascal_case, camel_case, and to full_case.

AhmadNader319 committed:
NameConventio class with methods to pascal_case, camel_case, and to full_case.

AhmadNader319 committed:
Update visitLocalVariableDeclaration for Storing camelCase replacements for local variables.

SIGMazer commented on FormattingVisitor.py:
```diff
@@ -75,29 +85,85 @@ def visitMethodDeclaration(self, ctx: JavaParser.MethodDeclarationContext):
+
+        return self.visitChildren(ctx)
+
+    def visitFormalParameter(self, ctx: JavaParser.FormalParameterContext):
+        if ctx.variableDeclaratorId():
+            param_name = ctx.variableDeclaratorId().getText()
+            camel_case = NameConventionHandler.to_camel_case(param_name)
+            self.parameter_replacements[param_name] = camel_case
+            self.rewriter.replaceRangeTokens(ctx.variableDeclaratorId().start, ctx.variableDeclaratorId().stop, camel_case)
+
+        return self.visitChildren(ctx)
+
+    def visitLocalVariableDeclaration(self, ctx: JavaParser.LocalVariableDeclarationContext):
+        if ctx.variableDeclarators():
+            for variable in ctx.variableDeclarators().variableDeclarator():
+                if variable.variableDeclaratorId():
+                    var_name = variable.variableDeclaratorId().getText()
+                    camel_case = NameConventionHandler.to_camel_case(var_name)
+                    self.local_variable_replacements[var_name] = camel_case
+                    self.rewriter.replaceRangeTokens(variable.variableDeclaratorId().start, variable.variableDeclaratorId().stop, camel_case)
+
+        return self.visitChildren(ctx)
+
+    def visitIdentifier(self, ctx: JavaParser.IdentifierContext):
```
we need to make the naming convention configurable check for `.java-format.json` in #8 or in master to see how we configure the identifiers

SIGMazer reviewed (changes requested):
remove unused files such as `formatting_changes.log` and `NameConversion.py` and resolve conflicts

AhmadNader319 committed:
Merge branch 'feat/name-convention' of https://github.com/0xCator/javalang-formatV2 into feat/name-convention

AhmadNader319 committed:
Initial: Implement name convention formatting based on configuration for methods and parameters

AhmadNader319 committed:
Appy name conversion using RegEx and add replacing functionalities

SIGMazer commented on NameConventionFormatterVisitor.py:
```diff
@@ -0,0 +1,96 @@
+    def visitClassDeclaration(self, ctx: JavaParser.ClassDeclarationContext):
+        class_name = ctx.identifier().getText()
+        print(f"Class name before: {class_name}")
+        if not re.match(self.config.naming_conventions['class'], class_name):
+            new_class_name = self.convert_to_pascal_case(class_name)
+            print(f"Class name after: {new_class_name}")
+            self.rewriter.replaceSingleToken(ctx.identifier().start, new_class_name)
+            self.replace_usage(class_name, new_class_name)
+        return self.visitChildren(ctx)
```
check for other config cases

AhmadNader319 replied:
I checked that.
Now all default configurations are loaded.

AhmadNader319 committed:
Load ConfigClass default configs with RegEx checking,
Enhancement: Cover more naming contexts

AhmadNader319 committed:
Merge branch 'master' into feat/name-convention

AhmadNader319 committed:
Added pattern transformer logic to enhance naming convention handling, applied naming conventions on the formatted code.

AhmadNader319 committed:
Merge branch 'feat/name-convention' of https://github.com/0xCator/javalang-formatV2 into feat/name-convention

NexusrexDev commented on FormattingVisitor.py:
```diff
@@ -1,13 +1,14 @@
-    def __init__(self, tokens, config: ConfigClass):
-        self.rewriter : TokenStreamRewriter = TokenStreamRewriter(tokens)
+    def __init__(self, rewriter, config: ConfigClass):
+        self.rewriter = rewriter
```
There's no need to use an external rewriter, please revert this class to its older version (from the latest commit in the `master` branch).

NexusrexDev commented on ConfigClass.py:
```diff
@@ -1,13 +1,9 @@
 
 class ConfigClass:
-    def __init__(self, config_path):
-        self.config_path = config_path
```
Removing this removed the loading aspect, when you need to try a different regex, please change it in the json file instead of modifying the class.

NexusrexDev commented on NameConventionFormatterVisitor.py:
Good work! Consider that we provide the option to use standard conventions such as camelcase and pascalcase, check `ErrorLogger.py` as a reference for its usage.
**EDIT**: The visitor applies the conventions to the imports, which will result in issues, fix that as well please.

NexusrexDev commented on formatting_changes.log:
Don't forget to remove the unused files.

NexusrexDev commented on testmain.py:
```diff
@@ -21,13 +24,39 @@ def parse_java_code(file_path):
+formatted_tree = formatted_parser.compilationUnit()
+
+
+name_rewriter = TokenStreamRewriter(formatted_tokens) 
+
+# Apply naming convention formatting
+name_formatter = NameConventionFormatterVisitor(name_rewriter, configs)
```
It'll be better to create the rewriter inside the class, akin to the way the FormattingVisitor used it initially.

NexusrexDev commented on testmain.py:
```diff
@@ -21,13 +24,39 @@ def parse_java_code(file_path):
+
+# Apply naming convention formatting
+name_formatter = NameConventionFormatterVisitor(name_rewriter, configs)
+name_formatter.visit(formatted_tree) 
+
+
+name_formatted_code = name_rewriter.getDefaultText()
+print("\nName Formatted Code:\n", name_formatted_code)  
```
It'll be far better to put that behind a single method for the ease of use. Again, FormattingVisitor's `get_formatted_code`.

SIGMazer commented on PatternTransformer.py:
revert the modification in this file it will make unwanted conflict in the feature

AhmadNader319 committed:
Added check_convention static method to return regex patterns for naming conventions. Use tokenStreamRewriter initially.
Parse the tree and return formatted code by get_formatted_code

AhmadNader319 commented:
In the last commit i did 
- Refactor NameConventionFormatterVisitor to enforce naming conventions
- Updated the constructor to initialize TokenStreamRewriter with tokens and added imports dictionary.
- Added check_convention static method to return regex patterns for naming conventions by calling StandardNamingConventoin enum class.
- Added _order_imports method to sort and merge imports based on configuration.
- Added get_formatted_code method to visit the parse tree and return formatted code.

AhmadNader319 committed:
Refactor naming convention checks to simplify regex pattern matching by removing the analyzer.

AhmadNader319 committed:
Merge branch 'master' into feat/name-convention

NexusrexDev commented:
Looks good to me, anything comes to mind before we get to merge @SIGMazer?

SIGMazer commented:
> Looks good to me, anything comes to mind before we get to merge @SIGMazer?

Let me check

SIGMazer reviewed (changes requested):
everything seam fine to me, but could you remove the `formatting_changes.log` and `.vscode` before you merge @AhmadNader319

AhmadNader319 committed:
Remove unused files.

AhmadNader319 committed:
Merge branch 'feat/name-convention' of https://github.com/0xCator/javalang-formatV2 into feat/name-convention

SIGMazer reviewed (approved):
@AhmadNader319 Nice work, you can merge it!
